File: RobotSimulation_MainClass/Robot.py
from Motor import Motor
from PID import PID
import math
"""
Класс робота - основной класс
При инициализации передать праметры робота:
    Радиус колёс
    Расстояние между колесaми - база
Также необходимо передавать начальные координаты и ориентацию робота
"""
PID_ITERATIONS = 200        # Колическо итераций для ПИД-регулятора
MOVE_ITERATIONS = 6000       # Количество итераций для функции Ляпунова
MOVE_ERROR = 0.0001              # Предельная ошибка регулирования фуекции Ляпунова
SIMULATION_DT = 0.01
class Robot:
    #Функция инициализации
    def __init__(self, x0, y0, angle0, wheelRadius, base):
        self.x = x0
        self.y = y0
        self.angle = angle0
        self.angle0 = angle0
        self.path = 0
        self.base = base
        self.wheeRadius = wheelRadius
        self.LeftMotor = Motor('L', SIMULATION_DT)
        self.RightMotor = Motor('R', SIMULATION_DT)

    #Движение робота
    def drive(self, Lvoltage, Rvoltage):
        self.LeftMotor.runMotor(Lvoltage)
        self.RightMotor.runMotor(Rvoltage)
        self.update()

    #Обновление информации и роботе
    def update(self):
        #Оновляем информацию о двигателях
        self.LeftMotor.update()
        self.RightMotor.update()

        #Запоминаем старые значения координат, пути и ориентации
        lastPath = self.path
        lastX = self.x
        lastY = self.y

        #Вычисляем новые значения
        self.path = (self.LeftMotor.theta + self.RightMotor.theta) * self.wheeRadius / 2  # вычисление пройденного пути
        self.angle = self.angle0 + (self.RightMotor.theta - self.LeftMotor.theta) * self.wheeRadius / self.base  # вычисление курсового угла
        self.angle = angleNormal180(self.angle)
        self.x = lastX + (self.path - lastPath) * math.cos(self.angle)  # вычисление координаты X
        self.y = lastY + (self.path - lastPath) * math.sin(self.angle)  # вычисление координаты Y

        #print('Information about robot`s odomeria was successfully updated!')

    #Поворот в указанную ориентацию при помощи ПИД-регулятора
    #Возвращает массив с историей регулирования для построения графика
    def rotateToAngle(self, purposeAngle):
        purposeAngle = angleNormal180(purposeAngle)
        Kp = 5
        Ki = 0
        Kd = 0
        pid = PID(Kp, Ki, Kd, SIMULATION_DT, purposeAngle, self.angle)
        for i in range(PID_ITERATIONS):
            voltage = pid.getControlValue(self.angle)
            self.drive(-voltage, voltage)
        return pid.getHistory()

    #Движение в указанную точку функцией Ляпунова
    def moveToPoint(self, targetX, targetY):
        moveHistoryX = []
        moveHistoryY = []
        for i in range(MOVE_ITERATIONS):
            moveHistoryX.append(self.x)
            moveHistoryY.append(self.y)
            distanceError = dist([self.x, self.y], [targetX, targetY])
            if distanceError < MOVE_ERROR:
                break
            bearing = -math.atan2(self.y - targetY, targetX - self.x)
            bearing = angleNormal180(bearing)
            courseAngle = bearing - self.angle
            control_l = 8 * math.cos(courseAngle) * math.tanh(distanceError) - 5 * courseAngle
            control_r = 8 * math.cos(courseAngle) * math.tanh(distanceError) + 5 * courseAngle
            self.drive(control_l, control_r)
        return moveHistoryX, moveHistoryY

#Расстояние между 2 точками
def dist(p1, p2):
    return ((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2) ** 0.5
# Нормализация значений угла в промежутке от 0 до 360
def angleNormal360(angle):
    angle %= math.pi * 2.0
    if angle < 0: angle += math.pi * 2.0
    return angle
#Нормализация угла в пределах [-180, 180]
def angleNormal180(angle):
    if angle == math.pi:
        return angle
    angle = (angle + math.pi) % (math.pi * 2.0)
    if (angle < 0.0):
        angle += math.pi * 2.0
    return angle - math.pi

File: ArduinoCode/Robot.cpp
#include "Robot.h"

Robot::Robot(double x0, double y0, double angle0, double wheelRadius, double base) {
  this->x = x0;
  this->y = y0;
  this->angle = angle0;
  this->angle0 = angle0;
  this->path = 0;
  this->base = base;
  this->wheelRadius = wheelRadius;
  this->LeftMotor = Motor('L', MOTOR_L_PIN_A, MOTOR_L_PIN_B, ENCODER_L_PIN_A, ENCODER_L_PIN_B);
  this->RightMotor = Motor('R', MOTOR_R_PIN_A, MOTOR_R_PIN_B, ENCODER_R_PIN_A, ENCODER_R_PIN_B);
}

Robot::Robot() {
  this->x = 0;
  this->y = 0;
  this->angle = 0;
  this->angle0 = 0;
  this->path = 0;
  this->base = 0;
  this->wheelRadius = 0;
  this->LeftMotor = Motor('L', MOTOR_L_PIN_A, MOTOR_L_PIN_B, ENCODER_L_PIN_A, ENCODER_L_PIN_B);
  this->RightMotor = Motor('R', MOTOR_R_PIN_A, MOTOR_R_PIN_B, ENCODER_R_PIN_A, ENCODER_R_PIN_B);
}

void Robot::TickL(){
  LeftMotor.Tick();
}
void Robot::TickR(){
  RightMotor.Tick();
}

void Robot::drive(double Lvoltage, double Rvoltage) {
  LeftMotor.runMotor(Lvoltage);
  RightMotor.runMotor(Rvoltage);
  delay(MOTOR_WORK_DELAY);
  Update();
}
void Robot::Stop() {
  drive(0, 0);
  delay(MOTOR_WORK_DELAY);
  Update();
}
//Обновление информации и роботе
void Robot::Update() {
  //Обновляем информацию о двигателях
  LeftMotor.Update();
  RightMotor.Update();

  //Запоминаем старые значения координат, пути и ориентации
  double lastPath = path;
  double  lastX = x;
  double lastY = y;

  //----------------------------Вычисляем новые значения---------------------------------
  // вычисление пройденного пути
  path = (LeftMotor.getTheta() + RightMotor.getTheta()) * wheelRadius / 2.0;
  // вычисление курсового угла
  angle = angle0 + (RightMotor.getTheta() - LeftMotor.getTheta()) * wheelRadius / base;
  angleNormal(angle);
  x = lastX + (path - lastPath) * cos(angle);  // вычисление координаты X
  y = lastY + (path - lastPath) * sin(angle); // вычисление координаты Y
}
//Поворот в указанную ориентацию при помощи ПИД - регулятора
//Возвращает массив с историей регулирования для построения графика
void Robot::rotateToAngle(double purposeAngle) {
  angleNormal(purposeAngle);
  double Kp = 20.0;
  double Ki = 0.0;
  double Kd = 0.0;
  PID pid(Kp, Ki, Kd, purposeAngle, angle);
  for (int i = 0; i < PID_ITERATIONS; i++) {
    double voltage = pid.getControlValue(angle);
    drive(-voltage, voltage);
  }
  Stop();
}

//Движение в указанную точку функцией Ляпунова
void Robot::moveToPoint(double targetX, double targetY) {
  for (int i = 0; i < MOVE_ITERATIONS; i++) {
    double distanceError = dist(x, y, targetX, targetY);
    if (distanceError < MOVE_ERROR)  break;
    double  bearing = atan2(targetY - y, targetX - x);
    double courseAngle = bearing - angle;
    angleNormal180(courseAngle);
    double control_l = 30.0 * cos(courseAngle) * tanh(distanceError) - 10.0 * courseAngle;
    double control_r = 30.0 * cos(courseAngle) * tanh(distanceError) + 10.0 * courseAngle;
    drive(control_l, control_r);
  }
  Stop();
}

//Расстояние между 2 точками
double dist(double x1, double y1, double x2, double y2) {
  return sqrt((x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1));
}
// Нормализация значений угла в промежутке от 0 до 360
void angleNormal360(double& angle) {
  angle = fmod(angle, PI * 2.0);
  if (angle < 0.0) {
    angle += PI * 2.0;
  }
}
// Нормализация угла в пределах[-180, 180]
void angleNormal180(double& angle) {
  if (angle == PI) return;
  angle = fmod(angle + PI, PI * 2.0);
  if (angle < 0.0) {
    angle += PI * 2.0;
  }
  angle -= PI;
}
//Нормализация углы тупым остатком от деления - эффективно для устронения проблемы перхода PID регулятора
void angleNormal(double& angle) {
  angle = fmod(angle, PI * 4.0);
}
